### ISSUE TITLE
Clang does not detect narrowing in aggregate (array) initialization

### ISSUE THREAD

**Issue by dangelog** [llvm/llvm-project#49020]
|  |  |
| --- | --- |
| Bugzilla Link | [49676](https://llvm.org/bz49676) |
| Version | trunk |
| OS | Linux |
| CC | @AaronBallman,@DougGregor,@zygoloid |

## Extended Description 
Testcase:
```cpp
    extern long double l;

    struct S {
        S(int) {}
    };

    S s[] = { l };
```

This is supposed to be ill-formed because of narrowing. Clang does not reject the code, nor it warns (as it usually does for narrowing).

The relevant rule is in [dcl.init.aggr]:

https://eel.is/c++draft/dcl.init.aggr#4.2.sentence-2

> Otherwise, the element is copy-initialized from the corresponding initializer-clause or is initialized with the brace-or-equal-initializer of the corresponding designated-initializer-clause.
> If that initializer is of the form assignment-expression or = assignment-expression and a narrowing conversion ([dcl.init.list]) is required to convert the expression, the program is ill-formed.

That rule has been changed in C++20 with designated initializers, but the same one was also there for C++17 and before, e.g. https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#3

For comparison, GCC accepts with no warnings, but MSVC warns about narrowing.

For reference, I've reported this very same bug against GCC here https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99625 .

**Comment by AaronBallman** [comment]
Clang's behavior is correct; that code should not issue a narrowing diagnostic. A slightly modified example should:
```
struct S {
  S(int) {}
};

S s[] = { { 1.0 } };
```
Without the extra curly braces around `1.0` (as in the original example), aggregate initialization is performed for the array object. Copy initialization is performed for each element of the array. The element is not an aggregate type (it has a user-provided constructor), so we hit this case: https://eel.is/c++draft/dcl.init#general-16.6.3 which does not diagnose a narrowing conversion.

With the extra curly braces, aggregate initialization is still performed for the array object, and copy initialization is performed for each element of the array. The element is still not an aggregate, but we hit this case https://eel.is/c++draft/dcl.init#general-16.1 and go through list initialization and hit https://eel.is/c++draft/dcl.init#list-3.7 which does diagnose a narrowing conversion.

**Comment by dangelog** [comment]
Hi,

The analysis is correct, but there's actually a core defect about this: https://github.com/cplusplus/CWG/issues/686 . 

I concur with the issue: I think [P3106R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3106r1.html) has _accidentally_ dropped the provisions for narrowing conversions from https://eel.is/c++draft/dcl.init.aggr#4.3 . With those provisions reinstated, the example as originally posted (single brace) should still be ill-formed. This was also what C++17 mandated in https://timsong-cpp.github.io/cppwp/n4659/dcl.init.aggr#3 .
